### PR TITLE
[SYCL][ESIMD][EMU] Queue/Event handling change for esimd_cpu

### PR DESCRIPTION
**Pull request by dongkyunahn-intel**
- Supporting only in-order queue for pi_esimd_cpu
- Event creation is changed to fail-safe one. Imported from pi_cuda.cpp

### CODE REVIEW

**Comment by romanovvlad** [comment]
> Including esimd_cpu in work-around for revised queue::wait() logic
same as level-zero, introduced in PR#4044
> Returning 'false' from ExecCGCommand:: producesPiEvent() predicate...

Could you please clarify why this is necessary? What is the problem with "dummy" events if all operations are completed during `piEnqueue*`?

Tagging @sergey-semenov

**Comment by dongkyunahn-intel** [comment]
Some ESIMD samples from intel/llvm-test-suite fails without this change - e.g. vadd_1d.

> What is the problem with "dummy" events if all operations are completed during piEnqueue*?

'DummyEvents' are still observed by `piEventsWait`. In order to completely eliminate events from `piEnqueueMemBufferRead` and `piEnqueueMemImageRead` done by CPU, the event returned from those PI_APIs must be `nullptr` (https://github.com/intel/llvm/pull/4011#discussion_r677815203). Even if `nullptr` is returned from those APIs, some ESIMD samples are still failing - e.g. accessor, PrefixSum.

Meanwhile, I have used ESIMD samples from intel/llvm-test-suite for enabling ESIMD_CPU support. So far, none of those samples has invoked `piEventCreate` function (this is why `piEventCreate` in PR#4011 is still empty). I would need an ESIMD sample actively using events for fully enabling this event handling.

**Comment by romanovvlad** [comment]
> 
> 
> Some ESIMD samples from intel/llvm-test-suite fails without this change - e.g. vadd_1d.
> 
> > What is the problem with "dummy" events if all operations are completed during piEnqueue*?
> 
> 'DummyEvents' are still observed by `piEventsWait`. In order to completely eliminate events from `piEnqueueMemBufferRead` and `piEnqueueMemImageRead` done by CPU, the event returned from those PI_APIs must be `nullptr` ([#4011 (comment)](https://github.com/intel/llvm/pull/4011#discussion_r677815203)). Even if `nullptr` is returned from those APIs, some ESIMD samples are still failing - e.g. accessor, PrefixSum.
> 
> Meanwhile, I have used ESIMD samples from intel/llvm-test-suite for enabling ESIMD_CPU support. So far, none of those samples has invoked `piEventCreate` function (this is why `piEventCreate` in PR#4011 is still empty). I would need an ESIMD sample actively using events for fully enabling this event handling.

Still not sure I understand. Can we return "dummy" events from piEnqueue* APIs and then just ignore them in all PI APIs that take events?

**Comment by dongkyunahn-intel** [comment]
> Still not sure I understand. Can we return "dummy" events from piEnqueue* APIs and then just ignore them in all PI APIs that take events?

`piEnqueue*` APIs and `piEventsWait` are implemented as your description. Some kernels are still failing without patch from this PR. I was investigating the failures after making kernels work but still cannot understand why.

`piEnqueueMemBufferRead`- https://github.com/intel/llvm/pull/4011/files#diff-d023d374bba88549a2b0637e35fb5d6d463b1fc885e86460bec7a563da17fa91R1198

`piEnqueueMemImageRead` - https://github.com/intel/llvm/pull/4011/files#diff-d023d374bba88549a2b0637e35fb5d6d463b1fc885e86460bec7a563da17fa91R1293

`piEventsWait` - https://github.com/intel/llvm/pull/4011/files#diff-d023d374bba88549a2b0637e35fb5d6d463b1fc885e86460bec7a563da17fa91R1078

**Comment by romanovvlad** [comment]
> 
> 
> > Still not sure I understand. Can we return "dummy" events from piEnqueue* APIs and then just ignore them in all PI APIs that take events?
> 
> `piEnqueue*` APIs and `piEventsWait` are implemented as your description. Some kernels are still failing without patch from this PR. I was investigating the failures after making kernels work but still cannot understand why.
> 
> `piEnqueueMemBufferRead`- https://github.com/intel/llvm/pull/4011/files#diff-d023d374bba88549a2b0637e35fb5d6d463b1fc885e86460bec7a563da17fa91R1198
> 
> `piEnqueueMemImageRead` - https://github.com/intel/llvm/pull/4011/files#diff-d023d374bba88549a2b0637e35fb5d6d463b1fc885e86460bec7a563da17fa91R1293
> 
> `piEventsWait` - https://github.com/intel/llvm/pull/4011/files#diff-d023d374bba88549a2b0637e35fb5d6d463b1fc885e86460bec7a563da17fa91R1078

I think we need to find the root cause before committing the patch.

**Comment by dongkyunahn-intel** [comment]
I think I found out solution. It has to be applied on top of the change set from PR#4011. As of August/19/2021, the PR is approved, but buildbot jobs fail due to missing packages required for online-building libva which is pre-requisite for CM. As soon as PR#4011 is merged into `origin/sycl` branch, I'll update this PR.

**Comment by romanovvlad** [comment]
@dongkyunahn-intel 

> Returning 'false' from ExecCGCommand:: producesPiEvent() predicate
for esimd_cpu as no effective event is created during esimd_cpu kernel
execution

Could you please tell if we can create a dummy event during esimd_cpu kernel execution?

**Comment by dongkyunahn-intel** [comment]
> Could you please tell if we can create a dummy event during esimd_cpu kernel execution?

Dummy event is already applied for `piEnqueueMemBufferRead` and `piEnqueueMemImageRead` in pi_esimd_cpu.cpp

https://github.com/intel/llvm/blob/bf7eb4cacca8b47208abf534f002955c1e9e9ec5/sycl/plugins/esimd_cpu/pi_esimd_cpu.cpp#L1084

https://github.com/intel/llvm/blob/bf7eb4cacca8b47208abf534f002955c1e9e9ec5/sycl/plugins/esimd_cpu/pi_esimd_cpu.cpp#L1214

https://github.com/intel/llvm/blob/bf7eb4cacca8b47208abf534f002955c1e9e9ec5/sycl/plugins/esimd_cpu/pi_esimd_cpu.cpp#L1309

**Comment by dongkyunahn-intel** [comment]
@romanovvlad, your suggestion worked with my fully working branch for ESIMD_CPU. After suggestion applied, only one file is changed.

Due to dependency on PR#4430, buildbots are failing during merge. I'll merge change sets from the PR to this PR when it is merged to `origin/sycl` branch.

**Comment by dongkyunahn-intel** [comment]
@romanovvlad / @kbobrovs ,

Can I apply change from this PR to PR#4430 (https://github.com/intel/llvm/pull/4430)? As this change set is limited to one file, and by applying the change, we could prevent possible failures in PR#4430 caused by not having changes from this PR.

**Comment by kbobrovs** [comment]
> @romanovvlad / @kbobrovs ,
> 
> Can I apply change from this PR to PR#4430 (#4430)? As this change set is limited to one file, and by applying the change, we could prevent possible failures in PR#4430 caused by not having changes from this PR.

Just curious - why not merge this PR first? I'm OK with uniting the two PRs.

**Comment by dongkyunahn-intel** [comment]
> Just curious - why not merge this PR first? I'm OK with uniting the two PRs.

It still has dependency on PR#4430 (4011 previously) as it changes some PI_API implementation.

File: sycl/plugins/esimd_cpu/pi_esimd_cpu.cpp
Comment (kbobrovs) [inline sycl/plugins/esimd_cpu/pi_esimd_cpu.cpp:739]: `piQueueFinish` must guarantee that all previously enqueued commands had completed. The fact that the queue is in-order does not guarantee that. In-order queue just means enqueued commands will execute in the enqueue order. Still, they can can execute asynchronously with the thread which enqueues the commands.

As far as I understand your current implementation, it is safe for `piQueueFinish` to be no-op just because *all* the commands like kernel execution or memory copy are blocking - enqueue of any command with ESIMD EMU plugin does not return until the command is actually finished.

If so, please change the comment.

File: sycl/plugins/esimd_cpu/pi_esimd_cpu.cpp
Comment (kbobrovs) [inline sycl/plugins/esimd_cpu/pi_esimd_cpu.cpp:1205]: As I understand your current implementation, with all commands being blocking, all events returned from enqueue operations are effectively in "completed" state once control is returned back to the caller of the enqueue operation.

- Given that I don't see the need in IsDummyEvent field.
- Second, I don't see the need in using `unique_ptr` here, as there is no chance RetEv will escape to any other thread
- Third, the above code is not needed, instead, the line 1218 should be '*Event = new _pi_event();' plus maybe setting the event to 'completed' state.

The same applies to other places where events are created.

Comment (dongkyunahn-intel) [reply sycl/plugins/esimd_cpu/pi_esimd_cpu.cpp:1205]: @romanovvlad suggested this implementation from CUDA PI implementation from recent phone call. Also, he was fine with this 'DummyEvent' approach.

Comment (kbobrovs) [reply sycl/plugins/esimd_cpu/pi_esimd_cpu.cpp:1205]: > @romanovvlad suggested this implementation from CUDA PI implementation from recent phone call. 

OK, now I see the point - unique_ptr is useful to automatically release the memory allocated by new _pi_even in case buf->CmBufferPtr->ReadSurface throws an exception. Disregard this my comment then.

> Also, he was fine with this 'DummyEvent' approach.
`DummyEvent` seems is redundant, as I explained above, even though harmless. So I don't mind leaving it as is for now.

File: sycl/plugins/esimd_cpu/pi_esimd_cpu.cpp
Comment (dongkyunahn-intel) [reply sycl/plugins/esimd_cpu/pi_esimd_cpu.cpp]: Changed the comment.

**Review by romanovvlad — APPROVED** [review]

**Comment by dongkyunahn-intel** [comment]
As PR#4430 is almost complete without change from this PR, I'll leave this and merge later.

**Review by kbobrovs — APPROVED** [review]

**Review by smaslov-intel — APPROVED** [review]

**Pull request merged**
